### PR TITLE
Joke customization

### PR DESCRIPTION
**Pull request by Vsumskiy**

### CODE REVIEW

File: src/app/pages/auth/auth-page/auth-page.component.ts [hunk 1]
```diff
@@ -12,28 +11,30 @@ import { AuthPropertiesEnum } from 'src/app/enums/enums';
...
   constructor(
     public authService: AuthService,
     private jokesDataService: JokesDataService,
     private jokeService: JokeService,
-    private route: ActivatedRoute,
     private fb: FormBuilder
   ) {}
 
   ngOnDestroy(): void {
     if (this.authService.authenticated) {
```
Comment (iyarymovych-reenbit) [inline src/app/pages/auth/auth-page/auth-page.component.ts:25]: In my mind, it's not a good place for fetching jokes from API after authentication.

File: src/app/pages/create-joke/create-joke-page/create-joke-page.component.ts [hunk 2]
```diff
@@ -1,18 +1,31 @@
...
+  ngOnDestroy(): void {
+    this.jokeService.setBufferJoke();
+  }
+
+  onNewJoke(joke: Joke): void {
+    this.jokeService.setBufferJoke(joke);
+    this.openSnackBar('Created!');
+  }
+
+  openSnackBar(message: string): void {
```
Comment (iyarymovych-reenbit) [inline src/app/pages/create-joke/create-joke-page/create-joke-page.component.ts:26]: I think this logic should live in a service.

File: src/app/pages/edit/edit-page/edit-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+import { Joke } from 'src/app/interfaces/interfaces';
+import { Component, OnInit, OnDestroy } from '@angular/core';
+import { MatSnackBar } from '@angular/material/snack-bar';
+
+@Component({
+  selector: 'app-edit-page',
+  templateUrl: './edit-page.component.html',
+  styleUrls: ['./edit-page.component.scss']
+})
+export class EditPageComponent implements OnInit, OnDestroy {
```
Comment (iyarymovych-reenbit) [inline src/app/pages/edit/edit-page/edit-page.component.ts:13]: `EditPageComponent` is not right name of the component.
What are you going to edit here? If joke, then `EditJokeComponent`.

File: src/app/pages/edit/edit-page/edit-page.component.html [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            Home
+        </a>
+        <h3>Jokes editor</h3>
+    </div>
+
+    <div class="container-inner">
+        <ng-container *ngIf="!loading; else loader"
+            class="cards">
+            <ng-container *ngIf="jokeService.getBufferJoke.length; else empty">
+                <app-create-edit-joke [joke]="jokeService.getBufferJoke[0]"
```
Comment (iyarymovych-reenbit) [inline src/app/pages/edit/edit-page/edit-page.component.html:16]: `JokeService` should be private in the component.
It's not a good idea to use it in a template.

Comment (iyarymovych-reenbit) [inline src/app/pages/edit/edit-page/edit-page.component.html:16]: We should talk regarding `app-create-edit-joke` component.

File: src/app/pages/edit/edit-page/edit-page.component.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    private snackBar: MatSnackBar,
+    public jokesDataService: JokesDataService
+  ) {}
+
+  ngOnInit(): void {
+    this.getCurrentJoke();
+  }
+
+  ngOnDestroy(): void {
+    this.jokeService.setBufferJoke();
```
Comment (iyarymovych-reenbit) [inline src/app/pages/edit/edit-page/edit-page.component.ts:27]: For what do you need this?

File: src/app/services/auth.service.ts [hunk 6]
```diff
@@ -44,9 +50,14 @@ export class AuthService {
     return JSON.parse(localStorage.getItem(this.localStorageKey));
   }
 
+  getAdminRole(user: UserInterface): boolean {
+    return user.roles.some(i => i === 'SuperAdmin');
```
Comment (iyarymovych-reenbit) [inline src/app/services/auth.service.ts:54]: 'SuperAdmin' value should live in some enum.

File: src/app/services/jokes-data.service.ts [hunk 7]
```diff
@@ -45,11 +51,35 @@ export class JokesDataService {
...
 
   removeFromDb(id: number | string): Observable<string> {
     return this.http.delete<string>(`${this.apiUrl}favorite/${id}`);
   }
+
+  deleteJoke(id: number | string): Observable<string> {
+    return this.http.delete<string>(this.apiUrl + id);
+  }
+
+  getByIdFromApi(id: string | number): Observable<Joke> {
```
Comment (iyarymovych-reenbit) [inline src/app/services/jokes-data.service.ts:82]: What are you getting by ID from API?

File: src/app/shared/components/create-edit-joke/create-edit-joke.component.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+import { JokesDataService } from 'src/app/services/jokes-data.service';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { Component, OnInit, EventEmitter, Output, Input } from '@angular/core';
+
+@Component({
+  selector: 'app-create-edit-joke',
+  templateUrl: './create-edit-joke.component.html',
+  styleUrls: ['./create-edit-joke.component.scss']
+})
+export class CreateEditJokeComponent implements OnInit {
```
Comment (iyarymovych-reenbit) [inline src/app/shared/components/create-edit-joke/create-edit-joke.component.ts:15]: I don't like this component. We need to talk regarding this.

File: src/app/pages/auth/auth-page/auth-page.component.ts (hunk 1 above)
Comment (Vsumskiy) [reply src/app/pages/auth/auth-page/auth-page.component.ts:25]: done

File: src/app/pages/create-joke/create-joke-page/create-joke-page.component.ts (hunk 2 above)
Comment (Vsumskiy) [reply src/app/pages/create-joke/create-joke-page/create-joke-page.component.ts:26]: ok

File: src/app/pages/edit/edit-page/edit-page.component.ts (hunk 3 above)
Comment (Vsumskiy) [reply src/app/pages/edit/edit-page/edit-page.component.ts:13]: renamed to EditJokeComponent

File: src/app/pages/edit/edit-page/edit-page.component.html (hunk 4 above)
Comment (Vsumskiy) [reply src/app/pages/edit/edit-page/edit-page.component.html:16]: done

File: src/app/pages/edit/edit-page/edit-page.component.ts (hunk 5 above)
Comment (Vsumskiy) [reply src/app/pages/edit/edit-page/edit-page.component.ts:27]: now i use BehaviorSubject

File: src/app/services/auth.service.ts (hunk 6 above)
Comment (Vsumskiy) [reply src/app/services/auth.service.ts:54]: done

File: src/app/services/jokes-data.service.ts (hunk 7 above)
Comment (Vsumskiy) [reply src/app/services/jokes-data.service.ts:82]: joke :)

File: src/app/shared/components/create-edit-joke/create-edit-joke.component.ts (hunk 8 above)
Comment (Vsumskiy) [reply src/app/shared/components/create-edit-joke/create-edit-joke.component.ts:15]: done

**Pull request merged**